modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/fun/IgniteSqlFunctions.java

```diff
@@ -16,22 +16,149 @@
+    private static class RangeTable implements ScannableTable {
+        /** Start of the range. */
+        private final Object rangeStart;
+
+        /** End of the range. */
+        private final Object rangeEnd;
+
+        /** Increment. */
+        private final Object increment;
+
+        /**
+         * Note: {@code Object} arguments required here due to:
+         * 1. {@code NULL} arguments need to be supported, so we can't use {@code long} arguments type.
+         * 2. {@code Integer} and other numeric classes can be converted to {@code long} type by java, but can't be
+         *      converted to {@code Long} type, so we can't use {@code Long} arguments type either.
+         * Instead, we accept {@code Object} arguments type and try to convert valid types to {@code long}.
+         */
+        RangeTable(Object rangeStart, Object rangeEnd, Object increment) {
+            this.rangeStart = rangeStart;
+            this.rangeEnd = rangeEnd;
+            this.increment = increment;
+        }
+
+        /** {@inheritDoc} */
+        @Override public RelDataType getRowType(RelDataTypeFactory typeFactory) {
+            return typeFactory.builder().add("X", SqlTypeName.BIGINT).build();
+        }
+
+        /** {@inheritDoc} */
+        @Override public Enumerable<@Nullable Object[]> scan(DataContext root) {
+            if (rangeStart == null || rangeEnd == null || increment == null)
+                return Linq4j.emptyEnumerable();
+
+            long rangeStart = convertToLongArg(this.rangeStart, "rangeStart");
+            long rangeEnd = convertToLongArg(this.rangeEnd, "rangeEnd");
+            long increment = convertToLongArg(this.increment, "increment");
+
+            if (increment == 0L)
+                throw new IllegalArgumentException("Increment can't be 0");
+
+            return new AbstractEnumerable<@Nullable Object[]>() {
+                @Override public Enumerator<@Nullable Object[]> enumerator() {
+                    return new Enumerator<Object[]>() {
+                        long cur = rangeStart - increment;
+
+                        @Override public Object[] current() {
+                            return new Object[] { cur };
+                        }
+
+                        @Override public boolean moveNext() {
+                            cur += increment;
+
+                            return increment > 0L ? cur <= rangeEnd : cur >= rangeEnd;
+                        }
+
+                        @Override public void reset() {
+                            cur = rangeStart - increment;
+                        }
+
+                        @Override public void close() {
+                            // No-op.
+                        }
+                    };
+                }
+            };
+        }
+
+        /** */
+        private long convertToLongArg(Object val, String name) {
+            if (val instanceof Byte || val instanceof Short || val instanceof Integer || val instanceof Long)
```

Comment:
I think we may support BigInteger and zero-scaled BigDecimal as well

Reply:
Yes, we can, but do we really need it? H2 `system_range` supports only `Long` arguments, DuckDB `range` also supports only `Long` arguments and this looks enough.